dhall/src/Dhall/Lint.hs

```diff
@@ -28,17 +28,18 @@ lint :: Expr s Import -> Expr t Import
 lint =
   Dhall.Optics.rewriteOf
     subExpressions
-    ( \e ->
-                removeLetInLet e
-        `mplus` removeUnusedBindings e
+    ( \e -> removeLetInLet e
         `mplus` optionalLitToSomeNone e
     )
+    . Dhall.Optics.rewriteOf subExpressions removeUnusedBindings
```

Comment:
Are you sure that you want to compose `rewriteOf`s instead of using `mplus`?  The reason I ask is that I believe using `mplus` ensures that we don't miss newly-exposed opportunities for rewrites.

For example, consider the generic case of:

```haskell
rewriteOf optic (\e -> f e `mplus` g e)
```

... versus:

```haskell
rewriteOf optic f . rewriteOf optic g
```

The latter approach means that if the rewrite named `f` exposes new opportunities for the rewrite named `g`, they will be missed, whereas the former approach using `mplus` would still work because it will keep trying either `f` or `g`.

Or, to use regular expression notation, the former approach using `mplus` is sort of like:

```
(f + g)*
```

... whereas the latter approach using function composition of `rewriteOf` would be like:

```
f*g*
```

... which is not equivalent.

Reply:
Yea, this is not quite the intention I had when I rewrote the module like this

Reply:
Oh wait, I see why they are done sequentially.  It's because `unfoldNestedLets` is the opposite of `removeLetInLet`

Is there any way to do this that doesn't require the `unfoldNestedLets` preprocessing pass?

Reply:
Yes. In fact I tried a direct fix first, i.e. modifying `removeUnusedBindings` to walk through all binders in a let block. But that turned out to get really quite complicated... 

I think this is by far the cleanest solution (and time complexity wise the unfolding/folding is no more than a constant factor). One thing that I might try is rewriting `lint` in a more explicit fashion, with comments at each step &ndash; the point-free style is a bit hard to read.

Reply:
Alright, then we'll keep it the way you have it